# 변수
## 메모리
- 변수(variable)나 참조(reference)의 개념을 명확히 이해하려면 메모리부터 이해해야함
- 컴퓨터는 1과 0으로 이루어진 데이터를 처리
    - #### 1과 0을 표현할 수 있는 데이터 단위 = 비트(bit)
        - 1비트는 0혹은 1
    - #### 비트가 여덟 개 모이면 바이트(byte)
        - 1바이트는 01011010과 같이 0과 1이 총 여덟 개로 구성되어 있다
    - #### 기가바이트(gigabyte)는 1,024 * 1,024 * 1,024 = 1,073,741,824바이트

### 32비트와 64비트의 의미
- 컴퓨터는 데이터 단위로 비트를 사용
- #### 데이터를 한 번에 몇 개 보낼 수 있는지를 나타내는 지표
    - 32비트와 64비트
        - 한 번에 보낼 수 있는 데이터 개수가 32비트면 32비트 컴퓨터
        - 한 번에 보낼 수 있는 데이터 개수가 64비트면 64비트 컴퓨터
- #### 32비트 컴퓨터는 메모리 주소를 32비트로 표현
- #### 64비트 컴퓨터는 메모리 주소를 64비트로 표현

### 메모리에 우편번호를 매긴다
- 메모리에 우편번호처럼 주소를 부여하는 것
    - #### 컴퓨터는 0과 1밖에 모르므로 컴퓨터에 주소를 알려 주려면 주소도 2진수로 나타내야 한다
    - #### 메모리 주소 한 개는 메모리에서 1바이트를 가리킨다
- 8비트 컴퓨터는 2진수 여덟 개로 메모리 주소를 표현
- 32비트 컴퓨터는 메모리 주소 하나를 나타내는데 2진수 서른두 자릿수를 사용
- #### 그러므로 32비트에서 나타낼 수 있는 주소 개수는 $2^{32}$(=4,294,967,296)개
    - #### 메모리 주소 한 개는 1바이트를 가리키므로 32비트로는 $2^{32}$(=4,294,967,296)바이트를 표현 할 수 있다
    - #### 다시 4,294,967,296 바이트는 4 x 1,024 x 1,024 x 1,024바이트로 풀어 쓸 수 있다
        - #### 1KB = 1,024, 1MB = 1,024 x 1,024 , 1GB = 1,024 x 1,024 x 1,024
        - 즉, 32비트로는 총 4GB 메모리를 가리킬 수 있다
        - 64비트 컴퓨터는 $2^{64}$(=18,446,744,073,709,551,616)바이트를 가리킬 수 있다

## 변수의 의미
- Variable(변수)란 데이터를 저장할 수 있는 메모리 공간을 의미
- #### 중요한 점, 변수가 단순한 이름이 아니라 메모리 공간 자체를 의미한다는 점
- 변수는 숫자와, 문자, 객체, 함수까지도 담을 수 있다
- #### 변수에 담긴 값이나 가리키는 대상은 언제라도 변경될 수 있다

# 정수
## 컴퓨터에서 수를 표현하는 방법
- 수를 표현하는 방법을 기수법(numeral system)이라고 한다
- 밑수를 정하면 밑수 개수만큼의 숫자(digit)를 사용해 수를 나타낼 수 있다

### 10진수
- 10진수는 수를 표현하는 데 숫자를 총 열개 사용
- 이때 밑수는 10이고 0부터 9까지 총 열 개의 숫자로 모든 수를 표현한다
    - 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 -> 총 10개의 숫자
    
### 2진수
- 2진수는 수를 표현하는 데 숫자 0과 1만 사용
- 컴퓨터가 인식할 수 있는 표현 방법
    - 0, 1 -> 총 2개의 숫자
    
### 16진수
- 16진수는 수를 표현하는 데 숫자를 총 열여섯 개 사용
- 0~9까지 숫자 + a~f 알파벳
    - 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, a, b, c, d, e, f -> 총 16개의 숫자

## 10진수를 2진수로
- 진수 변환 (10진수 -> 2진수)
    - 10진수 25를 2진수로 바꾸려면??
        - 1. 25를 2의 거듭제곱의 합으로 쪼갠다. 25에서 가장 가까운 2의 거듭제곱 수는 16 = $2^4$. 따라서 25에서 16을 빼면 9가 남는다
            - 25 -> 16 + 9
        - 2. 9에 가장 근접한 2의 거듭제곱 수는 8 = 2^3. 따라서 9에서 8을 빼면 1이 남는다
            - 25 -> 16 + 8 + 1 = $2^4$ + $2^3$ + $2^0$
        - 3. $2^2$이나 $2^1$처럼 중간에 빠진 지수는 0을 이용해 표현한다
            - 25 -> 1 X $2^4$ + 1 X $2^3$ + 0 X $2^2$ + 0 X $2^1$ + 1 X $2^0$
        - 4. 2의 거듭제곱 수를 제외하고 앞의 수 1과 0만 모아 나열한다
            - 1 X $2^4$ + 1 X $2^3$ + 0 X $2^2$ + 0 X $2^1$ + 1 X $2^0$ -> 11001
        - 25를 2진수로 변환하면 11001이다.

#### python code 로 위의 예시 확인

In [3]:
bin(25)

'0b11001'

## 2진수를 10진수로
- 진수 변환 (2진수 -> 10진수)
    - 11001을 10진수로 바꾸면??
        - 1. 11001 의 오른쪽 끝부터 0, 1, 2, 3, 4를 차례로 적는다. 이 수는 지수를 나타낸다
            - 1 = 4, 1 = 3 , 0 = 2 , 0 = 1 , 1 = 0
        - 2. 적어 둔 수를 지수로 활용하여 2의 거듭제곱 수의 합으로 표현
            - 1 X $2^4$ + 1 X $2^3$ + 0 X $2^2$ + 0 X $2^1$ + 1 X $2^0$
        - 3. 2의 수식을 계산한 결과 값이 10진수이다.
            - 16 + 8 + 0 + 0 + 1 = 25

## 16진수를 2진수로
- 2진수와 16진수 사이의 진수 변환
- #### 16진수 한 자리는 0부터 f까지 최대 10진수 15까지 표현할 수 있다
    - 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, a, b, c, d, e, f
    - 마지막 15를 2진수로 변환하면 1111이다 , f = 1111
    - 16진수 한 자리는 2진수 네 자릿수까지 표현할 수 있다
- 16진수 9를 2진수로 변환하면??
- #### 16진수를 2진수로 변환할 때 0 ~ 9 까지 변환하는 방법은 10진수에서 2진수로 변환할 때 쓰는 방법과 동일
    - 9 -> 8 + 1 = $2^3$ + $2^0$
    - $2^3$ + $2^0$ = 1 X $2^3$ + 0 X $2^2$ + 0 $2^1$ + 1 X $2^0$
    - 결괏값 = 0b1001
- #### a는 어떻게 변환할까??
    - a는 10진수로 10이다
    - 1. 10 -> 8 + 2 = $2^3$ + $2^1$
    - 2. $2^3$ + $2^1$ = 1 X $2^3$ + 0 X $2^2$ + 1 X $2^1$ + 0 X $2^0$
    - 3. 결괏값 = 0b1010

## 컴퓨터는 메모리 주소를 16진수로 나타낸다
- 8비트 컴퓨터의 메모리 주소가 2진수로 0010 1101이라면 2진수 네 자릿수를 16진수 한 자릿수로 표현할 수 있으므로 0x2d로 나타낼 수 있다
- #### 2진수로 표현하면 8비트를 표현하는 데 여덟 자릿수가 필요하지만, 16진수로 표현하면 두 자릿수로 간단히 나타낼 수 있어 가독성을 높일 수 있다
- #### 이러한 이유로 메모리 주소를 나타낼 때는 16진수를 사용한다
- #### 32비트 컴퓨터는 서른두 자릿수의 2진수 수가 아닌 여덟 자릿수의 16진수 수를 이용해 표현한다

## 양의 정수
- 컴퓨터는 정수(Integer)를 1바이트, 2바이트, 4바이트, 8바이트 등 다양한 크기로 저장
- 여기서는 1바이트 기준으로 설명
    - 정수에는 양수와 음수가 있으므로 부호를 나타내는 데 1비트를 사용
        - #### 맨 앞의 비트가 0이면 양수, 1이면 음수
            - 예시) 1. 25를 메모리에 저장할 때 맨 앞의 비트는 양수이므로 0이 된다
            - 2. 10진수 25를 2진수로 변환하면 11001이므로 나머지 비트를 0으로 채운다
            - 25가 메모리에 저장된 형태는 0001 1001
            - 0(양수)00(빈곳을 채우는 0)1 1001(25를 2진수로 변환) , 1바이트 = 비트가 8개이다
- #### 맨 앞의 비트가 부호를 나타낸다는 사실만 알고 있다면 10진수를 2진수로 변환한 것과 같다
- #### 1바이트로 나타낼 수 있는 수의 크기??
    - 정수에는 음수가 포함되고 맨 앞의 비트를 부호로 사용하므로 표현할 수 있는 양수의 범위는 절반으로 줄어든다
    - #### 즉, 1바이트로 나타낼 수 있는 수의 범위는  -128 ~ 127 이다
    - 하지만 음의 정수를 표현할 때는 방식이 많이 달라진다

## 음의 정수
- #### 컴퓨터는 음수를 보수 형태로 저장한다
### 보수의 개념
- 보수란 쉽게 말해 '보충해 주는 수'이다
    - 예를 들어 10진수에서 9의 보수를 구한다고 가정
        - 3의 9의 보수는 3을 더해 9가 되는 수인 6이다
        - 26의 9의 보수는 73이다
        - 어떤 수의 각 자릿수 수를 9에서 빼면 9의 보수를 구할 수 있다
    - #### 10의 보수
        - 3의 10의 보수는 3의 9의 보수인 6에 1을 더한 값인 7이다
        - 26의 10의 보수는 9의 보수인 73에 1을 더한 값인 74이다
    - #### 세 자리 수 9의 보수와 10의 보수
        - 123의 9의 보수와 10의 보수를 계산
            - 123의 각 자릿수 수를 9에서 빼면 876이므로 123의 9의 보수는 876
            - 123의 10의 보수는 123의 9의 보수인 876에 1을 더한 값인 877
### 2의 보수
- 2진수 1010의 1의 보수는 얼마??
    - 구하는 방법은 10진수 때와 같다. 10진수에서 9의 보수를 구할 때 9에서 각 자릿수 수를 뺐으니 1에서 각 자릿수 수를 빼면 된다
    - 결괏값은 0101. 각 자릿수의 1과 0을 반전한 결과
- 2진수 1010의 2의 보수는 얼마??
    - 1010의 1의 보수가 0101이므로 여기에 1을 더하면 된다
    - 즉, 1010의 2의 보수는 0110이다
- #### 2의 보수가 중요한 이유는 컴퓨터가 음수를 표현할 때 2의 보수를 사용하기 때문이다
### 음수의 표현
- #### 컴퓨터는 음수를 2의 보수로 표현한다
    - 예를 들어 음수 -4는 어떻게 표현??
        - 1. 4를 2진수 0000 0100으로 변환하고 1의 보수를 구한다
        - 2. 1에서 각 자릿수 수를 빼면 (모든 비트를 반전하면) 1의 보수인 1111 1011을 구할 수 있다
        - 3. 이 수에 1을 더해 2의 보수를 구하면 최종 결과는 1111 1100이다
        - #### 즉, 컴퓨터는 -4를 1111 1100으로 표현한다

#### -4라는 '정수'를 컴퓨터 메모리에 저장되는 '바이트' 형태로 표현하는 코드

In [4]:
(-4).to_bytes(1, byteorder='little', signed = True)

b'\xfc'

#### 코드 해석
- 첫 번째 인자는 몇 바이트로 나타낼지 지정한다. 지금가지 예제를 1바이트로 진행 했으므로 1로 지정
- 두 번째 인자는 byteorder(바이트 오더)이다.
- 마지막 인자인 signed는 양수와 음수를 모두 표현할지 아니면 양수만 표현할지 정하는 인수이다
- 출력 값은 16진수로 표현된다
    - 0xFC를 2진수로 변환하면 1111 1100이다

### 2의 보수로 표현하는 이유
- 첫째, 양수와 음수를 모두 양수처럼 저장한다고 가정
    - 0000 0000과 1000 0000은 +0과 -0이 된다
    - #### 즉, 0을 표현하는 두 가지 방법이 존재하게 된다
    - 이렇게 되면 컴퓨터 입장에서는 수 하나를 더 표현할 수 있는데 비트 하나를 낭비하는 셈이다
    - 또한 두 수를 비교할 때 CPU에서 뺄셈을 하는데 +0 과  -0을 비교하면 결과 값이 예상과 다르게 나온다
- 둘째, 컴퓨터에서 정수의 뺄셈 과정을 살펴보자
    - #### 덧셈은 단순히 두 수를 더하면 되지만, 뺄셈은 2의 보수 개념을 활용한다
    - #### 9-4를 계산할 떄 9에서 4를 빼는 게 아니라 9와 -4를 더한다
        - 10진수 9를 2진수로 표현 0000 1001 에서 10진수 -4를 2의 보수로 표현한 1111 1100을 더한다
        - 계산 결과를 보면 받아올림(carrying)이 발생하여 1 0000 0101이 나오는데 이때 받아올림 수(맨 앞 네모 안에 있는 1)은 버리면 된다
        - 최종 결과는 0000 0101로 5가 된다

# 실수
## 실수 연산의 함정
#### 실수 연산 예제

In [5]:
a = 0.01
result = 0.0

In [6]:
for i in range(100):
    result += a

In [7]:
result

1.0000000000000007

- a가 0.01이고 이를 100번 더했으니 당연히 1이 나와야 하는데 result 값이 다른다
- 1에 매우 가까운 수라고는 해도 1은 아니다

#### 앞의 코드와 계산 방식이 같은 예제

In [8]:
a = 0.015625
a

0.015625

In [9]:
result = 0.0

In [10]:
for i in range(100):
    result += a

In [11]:
result

1.5625

- #### 정확한 결과가 나왔다
    - 앞서 위의 예제는 다른 값이 나왔다
    - #### 이 결과를 이해하려면 컴퓨터가 실수를 표현하는 방법인 부동소수점이라는 개념을 이해해야 한다
## 부동소수점
- 컴퓨터는 IEEE(Institute of Electrical and Electronics Endineers, 전기전자기술협회)가 1985년에 제정한 ANSI/IEEE 754-1985라는 표준에 따라 실수(real number)를 표현한다
    - 이 표준에 따른 표현법을 부동소수점(floating-point)이라고 부른다
    - #### 부동소수점의 '부'는 부표(浮標)를 말할 때 쓰는 '부'로 둥둥 떠다닌다는 의미이다
- 실수 123.456을 여러 가지 방식으로 표현한 것
    - 1.23456 X $10^2$
    - 12.3456 X $10^1$
    - 1234.56 X $10^{-1}$
    - 12345.6 X $10^{-2}$
    - #### 소수점 위치를 보면 앞에 있기도 하고 뒤에 있기도 하다. 마치 소수점이 떠다니는 것처럼 보인다고 해서 이러한 "실수 표현 방식을 부동소수점"이라고 부른다.

## 단정도와 배정도
- 부동소수점에는 단정도 부동소수점과 배정도 부동소수점이 있다
    - 단정도(single -precision)는 실수를 32비트(4바이트)로 표현
        - 부호 1비트, 지수부 8비트, 가수부 23비트로 구성된다
    - 배정도(double-precision)는 실수를 64비트(8바이트)로 표현
        - 부호 1비트, 지수부 11비트, 가수부 52비트로 구성된다
- 배정도는 실수를 표현하는 데 사용하는 비트 수가 단정도 보다 두배 많은 만큼 정밀도가 높다
    - 파이썬은 배정도를 사용한다

#### python code

In [12]:
import sys
sys.float_info

sys.float_info(max=1.7976931348623157e+308, max_exp=1024, max_10_exp=308, min=2.2250738585072014e-308, min_exp=-1021, min_10_exp=-307, dig=15, mant_dig=53, epsilon=2.220446049250313e-16, radix=2, rounds=1)

- sys라는 모듈을 import한 다음 float_info를 입력
    - 최댓값, 최솟값, 정밀도 등을 확인 할 수 있다

In [13]:
sys.float_info.max

1.7976931348623157e+308

- 배정도 부동소수점으로 표현할 수 있는 가장 큰 수
    - 여기서 e는 지수를 뜻하는 exponent의 줄임말
    - 지수만 봐도 10의 308승이니 엄청나게 큰 수란 걸 알 수 있다

In [14]:
sys.float_info.min

2.2250738585072014e-308

- 배정도 부동소수점으로 표현할 수 있는 가장 작은 수
    - 10의 -308승이니 엄청나게 작은 수란 것 알 수 있다

### 그만큼 8바이트(64비트) 배정도 부동소수점을 사용하면 표현 범위가 엄청나게 넓어진다
- 2.2250738585072014 X $10^{308}$

## 1바이트 실수 자료형 설계하기
- 실수 표현 방식을 쉽게 이해할 수 있도록 1바이트 실수 자료형을 직접 만들어 보자
    - 다음은 실수 자료형을 표현한 수식이다
    - $± 1.max$ X $2^{exp-bias}$
    - 1.man은 가수(mantissa/fraction), 2는 밑수, exp-bias는 지수(exponent)를 의미한다
- #### 위의 식을 이용하여 7.75라는 10진수 실수를 1바이트 부동소수점으로 표현해 보자
    - 이 식은 2진수에 대한 식이므로 우선 7.75라는 10진수를 2진수 실수로 바꾸는 작업부터 해야 한다

### 10진수 실수를 2진수 실수로 바꾸기
- 10진수 실수를 2진수로 바꾸는 방법
    - 정수를 바꿨던 방법과 크게 다르지 않다
    - #### 여기서 $2^{-1}$은 음수가 아니라 우리가 흔히 분수라고 부르는 1/2이다. $2^{-2}$은 1/4다
    - #### 그러므로 다음 수식처럼 쪼갠 다음 변환을 하면 111.11이 된다
        - 7.75 = 4 + 3 + 1 + 0.5 + 0.25
        - = $2^2$ + $2^1$ + $2^0$ + $2^{-1}$ + $2^{-2}$
        - = 111.11

### 정규화
- #### 정규화(normalization)란 소수점 왼쪽에 위치한 가수 부분을 밑수보다 작은 자연수가 되도록 만드는 것이다
    - 예를 들어 10진수 567.89를 정규화하면 소수점 왼쪽에 위치한 가수 부분이 밑수 10보다 작은 자연수 5가 되어 5.6789 X $10^2$이 된다
- 2진수의 밑수는 2이므로 2보다 작은 자연수는 1밖에 없다. 따라서 소수점 왼쪽의 가수 부분은 항상 1이 된다
    - 111.11 = 1.1111 X $2^2$

### 메모리 구조
- #### 정규화된 부동소수점 수 1.1111 X $2^2$을 앞의 수식과 비교해 보면 man은 1111이고 exp-bias는 2이다
- 이제 1바이트의 메모리 구조를 정하고 man과 exp 값만 저장하면 설계가 끝난다
    - #### 이때 지수부와 가수부에 할당하는 비트 수에 따라 표현하는 범위와 정밀도가 결정된다
- 1바이트 부동소수점의 구성
    - 0 0000 000
    - 0(부호) 0000(지수부) 000(가수부)
    - 실수 역시 정수와 마찬가지로 첫 번째 비트는 부호를 나타낸다
        - 0이면 양수, 1이면 음수
    - 가운데 4비트는 지수부로 exp 값을 저장
    - 맨 뒤 3비트는 가수부로 man 값을 저장
    - bias는 지수의 부호를 결정하는 데 쓴다
    - #### 부동소수점의 지수부에는 부호 비트가 없으며 0 ~ 15의 양수만 나타낼 수 있다
    - #### 음수 지수도 필요하다
        - 음수를 사용하려면 bias를 7로 두고 지수부(exp)에서 bias를 뺀 값을 실제 지수로 사용한다
        - bias는 $2^{n-1}-1$ 식에 지수부의 비트 수인 4를 대입하면 구 할 수 있다
- 정규화된 수 1.1111 X $2^2$에서 실제 지수는 2이다
    - 실제 지수가 2라는 의미는 exp-bias가 2라는 것이다
    - 지수부의 비트 수 4를 식($2{n-1}-1$)에 대입해 얻은 bias값이 7이므로 실제로 부동소수점의 지수부에 나타나는 값 exp는 9가 된다
- #### 1바이트 구성
    - 부호 -> 0
    - 지수부 -> 1001
    - 가수부 -> 1111
- #### 가수부는 3비트만 할당되는데 위의 값은 1111이다. 이런 경우에는 뒷자리 1을 생략한다
    - 즉, 가수부는 111이 된다
- #### 0 1001 111 = 0100 1111 = 0x4f
- #### 실수 7.75는 1바이트 부동소수점으로 나타내면 0x4f이다

### 1바이트 부동소수점의 표현 범위
- 위에서 설계한 1바이트 부동소수점으로 표현할 수 있는 가장 작은 수와 가장 큰 수는 다음과 같다
    - 표현할 수 있는 가장 작은 수(지수부가 0001일 때)
        - 1.000 X $2{-6}$ = 0.015625
    - 표현할 수 있는 가장 큰 수(지수부가 1110일 때)
        - 1.111 X $2^{7}$ = 240
- 1바이트지만 굉장히 작은 수부터 큰 수까지 폭넓게 표현할 수 있다는 것을 알 수 있다

### 1바이트 부동소수점의 정밀도
- 변환 과정에서 가수부를 담을 공간이 부족해 가수부에 들어갈 데이터인 1111에서 맨 뒤에 있는 1을 누락했다
    - #### 이렇게 되면 0x4f는 7.75라는 실수를 완벽하게 표현하지 못한다
    - 1.111 X $2{^2}$ = 1 X $2^{2}$ + 1 X $2^{1}$ + 1 X $2^{0}$ + 1 X $2^{-1}$ = 7.5
    - #### 0.25만큼 차이가 나므로 정밀도가 그만큼 떨어진다

## 정밀도에 대한 고찰
### 엡실론
- #### 실수 자료형에서 엡실론(epsilon)이란 1.0과 그 다음으로 표현 가능한 수(representable float)사이의 차이를 말한다

#### python code

In [15]:
import sys
sys.float_info.epsilon

2.220446049250313e-16

- 위 코드는 sys 모듈의 float_info에 있는 epsilon을 출력한 결과이다
    - 파이썬이 사용하는 배정도(double)의 가수부가 52비트
    - 1.0을 배정도에 맞춰 표현하면 다음과 같다
        - 1.000....000(0: 52개) X $2^{0}$
    - 배정도에서 1.0 다음으로 표현할 수 있는 수(representable float)는 다음과 같다
        - 1.000...0001(0: 51개, 1: 마지막 비트) X $2^{0}$
    - 따라서 두 수의 차이는 다음과 같다
        - 0.000....0001(0: 51개, 1: 마지막 비트) X $2^{0}$
    - 이 수를 10진수로 바꾸면 엡실론 값이 나온다
        - 2.220446049250313 X $10^{-16}$

### 엡실론과 정밀도
- 어떤 실수가 있을 때 엡실론을 이용하면 그 실수 다음에 표현할 수 있는 수를 알아낼 수 있다
    - 예를 들어 배정도 실수 9.25를 부동소수점 방식으로 표현하면 1.00101 X $2^3$이다
    - 이 식에서 지수 부분만 떼어 내 엡실론을 곱하면 이 실수와 다음 표현 가능한 수 사이의 차이를 구할 수 있다

#### python code

In [16]:
import sys
ep = sys.float_info.epsilon

In [17]:
a = 9.25

In [19]:
#1
diff = (2**3)*ep
diff

1.7763568394002505e-15

In [20]:
#2
b = a + diff
b

9.250000000000002

- #### #1에서 diff는 지수 부분인 $2^3$에 엡실론을 곱한 값으로 9.25와 그 다음 표현 가능한 수 사이의 차이이다
- #### #2에서 b는 a에 diff를 더했으므로 9.25 다음에 표현 가능한 수를 나타낸다

#### 9.25에 diff보다 작은 값을 더하면 어떤 일이 일어날까??

In [21]:
a

9.25

In [22]:
half_diff = diff/2
half_diff

8.881784197001252e-16

In [23]:
c = a + half_diff

In [24]:
a == c

True

- half_diff는 diff 값의 반절이다. half_diff 역시 실수를 나눠서 얻은 값이니 실수이다
- a와 half_diff를 더해 c에 대입한다. 실수와 실수를 더했으므로 c는 반드시 a보다 커야한다
- #### 그런데 a값과 c값이 같다고 나온다 어떻게 된 일일까??

- #### a 값과 c 값이 같은 이유는 c에 더한 half_diff 값이 diff보다 작기 때문이다
    - diff는 9.25와 그다음 표현 가능한 수 사이의 차이이므로 9.25에 diff보다 작은 값을 더한 수를 부동소수점 방식에서는 표현할 수 없다
    - 달리 표현하면 정밀도가 떨어진다는 말이다

#### 부동소수점의 정밀도에 관한 예
- a 는 1.000....000(0 : 52개) X $2^{53}$이다
    - a와 a 다음에 표현 가능한 수 사이의 차이는 $2^{53}$에 엡실론을 곱해 구할 수 있으며 그 값은 2.0이다
    - 그러므로 a에 1.0을 더한 b는 원래 의도한 값을 표현하지 못하고 a와 같은 값을 갖게 된다
    - #### 1.0의 차이조차 표현할 수 없는 정밀도이다

#### python code

In [28]:
a = (2.0)**53
a

9007199254740992.0

In [29]:
b = a + 1.0

In [30]:
a == b

True